src/semant.ml

```diff
@@ -1,112 +1,194 @@
-let missing_id_err = "variable does not exist"
+let duplicate_func_err = "function name already exists"
+let duplicate_param_name_err = "function parameters must have unique names"
 let invalid_assignment_err = "variable type does not match assigned value"
+let invalid_bop_args_err = "invalid arguments for binary operator"
+let invalid_cond_loop_err = "conditional loop must take in a boolean expression"
 let invalid_if_err = "if must take in a boolean expression"
 let invalid_iter_loop_err = "iterative loop must take in numbers"
-let invalid_cond_loop_err = "conditional loop must take in a boolean expression"
-let duplicate_func_err = "function name already exists"
-let missing_func_err = "function does not exist"
-let invalid_bop_args_err = "invalid arguments for binary operator"
-let mismatched_bop_args_err = "mismatched arguments for binary operator"
 let invalid_unop_args_err = "invalid argument for unary operator"
-let none_return_err = "function with non-none return type does not return anything"
 let mismatched_func_args_err = "mismatched arguments for function call"
-let return_in_global_err = "cannot return outside a function" (* TODO: implement *)
-let return_in_none_err = "function that returns none cannot have return statement"
+let mismatched_bop_args_err = "mismatched arguments for binary operator"
 let mismatched_return_err = "incorrect function return type"
+let missing_func_err = "function does not exist"
+let missing_id_err = "variable does not exist"
 let missing_return_err = "missing return statement"
+let none_return_err = "function with non-none return type does not return anything"
+let nonguaranteed_return_err = "function is not guaranteed to return"
+let return_in_global_err = "cannot return outside a function"
+let return_in_none_err = "function that returns none cannot have return statement"
 let unimplemented_err = "unimplemented"
 
 let check (prog: program): sprogram =
-  let name_exists tables name =
-    let (ids, funcs) = tables in (StringMap.mem name ids || StringMap.mem name funcs)
-  in let rec find_id name ids_table = match ids_table with
-      [] -> raise (Failure missing_id_err)
-    | curr_ids::outer_ids -> if StringMap.mem name curr_ids then StringMap.find name curr_ids else find_id name outer_ids
-
-  in let rec check_expr tbs = function
-      NumberLit n -> (Number, SNumberLit n)
-    | BoolLit b -> (Bool, SBoolLit b)
-    | StringLit s -> (String, SStringLit s)
-    | CharLit c -> (Char, SCharLit c)
-    | Id id -> (find_id id (fst tbs), SId id)
-    | Binop (e1, op, e2) -> let (type1, sexpr1) = check_expr tbs e1 in let (type2, sexpr2) = check_expr tbs e2 in
-        let final_type = if type1 = type2 then (
-          match op with
-              (Plus | Minus | Times | IntDiv | Div | Mod) when type1 = Number -> Number
-            | (Eq | Neq | Less | Leq | Greater | Geq) when type1 = Number -> Bool
-            | (And | Or) when type1 = Bool -> Bool
+  let default_capacity = 10 in
```

Comment:
what's this?

Reply:
just whenever I create a new hashtable, I need to pass in an initial capacity. any number works, I just chose 10.